src/os/bluestore/BlueStore.cc

```diff
@@ -14590,18 +14591,20 @@ void BlueStore::_do_write_small(
     o->extent_map.punch_hole(c, offset, length, &wctx->old_extents);
 
     // Zero detection -- small block
-    if (!bl.is_zero()) {
+    if (!bl.is_zero() || !cct->_conf->bluestore_zero_block_detection) {
```

Comment:
Check if `bluestore_zero_block_detection` is not set first -- that way a potentially costly bufferlist `mem_is_zero` test would be avoided.